src/components/TracingVisualization/TracingVisualization.tsx

```diff
@@ -1,33 +1,51 @@
-import { IApolloServerTracing } from "@/types"
-import { TracingVisualizationRow } from "."
+import { IApolloServerTracing, Maybe } from "@/types"
+import { TracingVisualizationRow, useTracingVirtualization } from "."
 
 interface ITracingVisualizationProps {
-  tracing?: IApolloServerTracing
+  tracing: Maybe<IApolloServerTracing>;
 }
 
 export const TracingVisualization = (props: ITracingVisualizationProps) => {
   const { tracing } = props
   const totalTimeNs = tracing?.duration || 0
+  const { ref, resolvers, totalSize, virtualItems, height } = useTracingVirtualization(tracing)
 
   return (
-    <>
-      <TracingVisualizationRow
-        type="total"
-        name="Total"
-        total={totalTimeNs}
-        duration={totalTimeNs}
-      />
-
-      {tracing?.execution.resolvers.map((a) => (
+    <div
+      ref={ref}
+      className="relative h-full overflow-y-auto p-4"
+    >
+      <div
+        className="relative w-full"
+        style={{
+          height: `${totalSize}px`,
+        }}
+      >
         <TracingVisualizationRow
-          key={a.path.join(".")}
-          type={a.parentType}
-          name={a.path.join(".")}
+          type="total"
+          name="Total"
           total={totalTimeNs}
-          offset={a.startOffset}
-          duration={a.duration}
+          duration={totalTimeNs}
         />
-      ))}
-    </>
+
+        {virtualItems.map(({ key, index, size, start }) => (
+          <TracingVisualizationRow
+            key={key}
+            type={resolvers[index].parentType}
+            name={resolvers[index].path.join(".")}
+            total={totalTimeNs}
+            offset={resolvers[index].startOffset}
+            duration={resolvers[index].duration}
+            style={{
+              position: 'absolute',
+              top: 0,
+              left: 0,
+              height: `${size}px`,
```

Comment:
only height and transform are dynamic here. Could we instead just pass these as props and render the style internally.

Also could we switch `position: 'absolute', top: 0, left: 0` to tailwind classnames to remain consistent with the rest of the app?